File: matlab/LearnMPC_GPs/detransform_data.m
function [Data] = detransform_data(Data_transformed,transform_info)


if strcmp('none',transform_info.method) % no transformation
    Data = Data_transformed;

end

if strcmp('center',transform_info.method) % zero mean
    sz = size(Data_transformed);
    Data = Data_transformed + repmat(transform_info.mean_data,sz(1),1);

end

if strcmp('zscore',transform_info.method) % data have zero mean and std of 1
    sz = size(Data_transformed);
    mean_data = transform_info.mean_data;
    std_data = transform_info.std_data;

    mean_data_mat = repmat(mean_data,sz(1),1);
    std_data_mat = repmat(std_data,sz(1),1);

    Data = Data_transformed.*std_data_mat + mean_data_mat;

end

if strcmp('rescale',transform_info.method)
    sz = size(Data_transformed);
    min_data = repmat(transform_info.min_data,sz(1),1);
    max_data =  repmat(transform_info.max_data,sz(1),1);
    a = transform_info.lb;
    b = transform_info.ub;

    Data =( (Data_transformed - a)./(b-a) ).*(max_data-min_data)+min_data;

end

end

File: python/src/GP_learning/detransform_data.py
# -*- coding: utf-8 -*-
"""
Created on Wed Nov 27 16:16:01 2024

@author: rose
"""

import numpy as np

def detransform_data(Data_transformed, transform_info):
    """
    Reverses the transformation applied to the data based on the provided transform_info.
    
    Parameters:
    Data_transformed (numpy.ndarray): The transformed data to be reversed.
    transform_info (dict): A dictionary containing the information needed to reverse the transformation.
    
    Returns:
    Data (numpy.ndarray): The original data before transformation.
    """
    
    if transform_info['method'] == 'none':  # No transformation
        Data = Data_transformed
    
    elif transform_info['method'] == 'center':  # Reversing centering (zero mean)
        Data = Data_transformed + transform_info['mean_data']
    
    elif transform_info['method'] == 'zscore':  # Reversing z-score normalization
        mean_data = transform_info['mean_data']
        std_data = transform_info['std_data']
        Data = Data_transformed * std_data + mean_data
    
    elif transform_info['method'] == 'rescale':  # Reversing rescaling
        min_data = transform_info['min_data']
        max_data = transform_info['max_data']
        a = transform_info['lb']
        b = transform_info['ub']
        Data = ((Data_transformed - a) / (b - a)) * (max_data - min_data) + min_data
    
    return Data